File: 6_parent_child.py
import os

MAX_SIZE = 100

def main():
    # Create two pipes: one for parent-to-child (p2c) and one for child-to-parent (c2p)
    p2c = os.pipe()  # Parent-to-Child
    c2p = os.pipe()  # Child-to-Parent

    # Fork the process
    pid = os.fork()

    if pid > 0:  # Parent process
        os.close(p2c[0])  # Close read end of p2c
        os.close(c2p[1])  # Close write end of c2p

        # Parent sends a message to the child
        msg = input("Parent, enter your message: ")
        os.write(p2c[1], msg.encode())  # Send message to the child
        os.close(p2c[1])  # Close write end after sending

        # Parent reads the child's response
        reply = os.read(c2p[0], MAX_SIZE).decode()
        print("Parent received:", reply)
        os.close(c2p[0])  # Close read end after receiving

    else:  # Child process
        os.close(p2c[1])  # Close write end of p2c
        os.close(c2p[0])  # Close read end of c2p

        # Child reads the parent's message
        reply = os.read(p2c[0], MAX_SIZE).decode()
        print("Child received:", reply)
        os.close(p2c[0])  # Close read end after receiving

        # Child sends a response back to the parent
        msg = input("Child, enter your response: ")
        os.write(c2p[1], msg.encode())  # Send response to the parent
        os.close(c2p[1])  # Close write end after sending

if __name__ == "__main__":
    main()

File: 6_parent_child.c
#include <stdio.h>
#include <unistd.h>
#include <string.h>

#define MAX_SIZE 100

int main() {
    int p2c[2], c2p[2];
    pid_t pid;
    char msg[MAX_SIZE], reply[MAX_SIZE];

    if (pipe(p2c) == -1 || pipe(c2p) == -1 || (pid = fork()) < 0) {
        perror("Error");
        return 1;
    }

    if (pid) {  // Parent
        close(p2c[0]); close(c2p[1]);
        printf("Parent, enter your message: ");
        fgets(msg, MAX_SIZE, stdin);  // Take input from parent
        msg[strcspn(msg, "\n")] = 0;  // Remove the newline character
        write(p2c[1], msg, strlen(msg) + 1);
        read(c2p[0], reply, MAX_SIZE);
        printf("Parent received: %s\n", reply);
    } else {  // Child
        close(p2c[1]); close(c2p[0]);
        read(p2c[0], reply, MAX_SIZE);
        printf("Child received: %s\n", reply);
        printf("Child, enter your response: ");
        fgets(msg, MAX_SIZE, stdin);  // Take input from child
        msg[strcspn(msg, "\n")] = 0;  // Remove the newline character
        write(c2p[1], msg, strlen(msg) + 1);
    }

    return 0;
}
